mountpoint-s3-crt/src/s3/endpoint_resolver.rs

```diff
@@ -0,0 +1,263 @@
+//! Endpoint resolver helps determine which S3 endpoint to use for a given configuration.
+
+use mountpoint_s3_crt_sys::*;
+use std::{
+    ffi::OsStr,
+    os::unix::prelude::OsStrExt,
+    ptr::{self, NonNull},
+};
+
+use crate::{
+    aws_byte_cursor_as_slice,
+    common::{allocator::Allocator, error::Error, uri::Uri},
+    CrtError, ToAwsByteCursor,
+};
+
+use super::s3_library_init;
+
+/// [RuleEngine] to resolve endpoint with given [RequestContext] and ruleset
+#[derive(Debug)]
+pub struct RuleEngine {
+    inner: NonNull<aws_endpoints_rule_engine>,
+}
+
+impl RuleEngine {
+    /// Creates a new endpoint [RuleEngine].
+    pub fn new(allocator: &Allocator) -> Result<Self, Error> {
+        s3_library_init(allocator);
+        // SAFETY: `allocator.inner` is a valid aws_allocator and and we check the return is non-null.
+        // SAFETY: aws_s3_endpoint_resolver_new will acquire a reference to keep it alive after this function call, so it's safe to return an owned version to it.
+        let inner = unsafe { aws_s3_endpoint_resolver_new(allocator.inner.as_ptr()).ok_or_last_error()? };
+        Ok(Self { inner })
+    }
+
+    /// Resolve the endpoint with given [RequestContext] and ruleset.
+    pub fn resolve(&self, context: RequestContext) -> Result<ResolvedEndpoint, Error> {
+        // SAFETY: `aws_endpoints_rule_engine_resolve` ensures that it returns a non null `aws_endpoints_resolved_endpoint`
+        // or it will return an error. So, the out_endpoint is valid and non null.
+        unsafe {
+            let mut out_endpoint: *mut aws_endpoints_resolved_endpoint = ptr::null_mut();
+            aws_endpoints_rule_engine_resolve(self.inner.as_ptr(), context.inner.as_ptr(), &mut out_endpoint)
+                .ok_or_last_error()?;
+            match aws_endpoints_resolved_endpoint_get_type(out_endpoint) {
+                aws_endpoints_resolved_endpoint_type::AWS_ENDPOINTS_RESOLVED_ERROR => Err(Error::last_error()),
+                _ => {
+                    let out_endpoint = out_endpoint.ok_or_last_error()?;
+                    Ok(ResolvedEndpoint { inner: out_endpoint })
+                }
```

Comment:
Still need to address this wildcard with an actual match of `AWS_ENDPOINTS_RESOLVED_ENDPOINT`.

Since this Enum is marked non_exhaustive, then maybe we can either return a generic error for something unexpected or panic.